### ISSUE TITLE
Data Flywheel: Machine Learning Foundation for Intelligent Spatial Organization

### ISSUE THREAD

**Issue by lukehanner** [modryn-studio/ProjectLoom#12]
# Data Flywheel: Machine Learning Foundation

## Overview

Implement the foundational data collection and ML infrastructure that creates ProjectLoom's competitive moat: a product that gets smarter with every user interaction. This transforms ProjectLoom from a static canvas tool into an intelligent spatial thinking partner.

## The Strategic Advantage

**Generic Chat Tools**: Same experience for everyone
**ProjectLoom with Flywheel**: Gets better with YOUR usage patterns

### The Flywheel Loop
1. User organizes conversations spatially
2. ML learns user's organization logic
3. System suggests card positions matching user's thinking
4. User accepts/rejects suggestions (more training data)
5. Accuracy improves → user saves time → uses tool more
6. More usage → more data → smarter suggestions
7. **Loop accelerates**

### 1. Auto-Organization Intelligence
**Training Data Collection:**
- Card creation position (x, y) with content type
- Card movement patterns (spatial grouping)
- Branch creation points in conversations

**Smart Features:**
- AI-suggested optimal card placement
- Auto-grouping of related cards
- Predictive branching recommendations

### 2. Context Inheritance Optimization
**Training Data Collection:**
- Manual message selection patterns
- Merge node creation patterns
- Context window adjustments per conversation type

**Smart Features:**
- Auto-suggest which messages to inherit
- Smart merge candidate recommendations
- Optimized token usage based on patterns

### 3. Conversation Synthesis
**Training Data Collection:**
- Summary creation timing and content
- Multi-thread merge patterns
- Cross-card reference patterns

**Smart Features:**
- Better auto-summaries of conversation threads
- Intelligent merge suggestions
- "You might want to reference this" predictions

### Data Collection Schema

```typescript
// Core telemetry types
interface UserAction {
  id: string
  userId: string
  timestamp: Date
  sessionId: string
  type: UserActionType
  metadata: ActionMetadata
}

type UserActionType = 
  | 'card_created'
  | 'card_moved'
  | 'card_deleted'
  | 'branch_created'
  | 'context_selected'
  | 'merge_created'
  | 'summary_generated'
  | 'card_expanded'
  | 'conversation_switched'

interface ActionMetadata {
  // Card operations
  cardId?: string
  cardPosition?: { x: number, y: number }
  cardSize?: { width: number, height: number }
  contentType?: 'conversation' | 'merge' | 'summary'
  
  // Context operations
  selectedMessages?: string[]
  contextTokenCount?: number
  inheritanceStrategy?: 'full' | 'partial' | 'none'
  
  // Spatial operations
  nearbyCards?: string[] // Cards within 500px
  canvasZoom?: number
  viewportCenter?: { x: number, y: number }
  
  // Conversation operations
  branchPoint?: number // Message index where branch occurred
  mergeSourceCards?: string[]
  conversationLength?: number
  
  // Performance
  actionDuration?: number // For UX optimization
}
```

### Storage Strategy

**Phase 2 (Local-First):**
```typescript
// IndexedDB for client-side storage
interface TelemetryStore {
  actions: UserAction[]
  aggregates: {
    cardPositionHeatmap: Map<string, number>
    commonBranchPoints: number[]
    averageContextSize: number
  }
}
```

**Phase 3+ (Optional Cloud Sync):**
- Anonymized aggregation for global insights
- User-specific model training
- Cross-device pattern sync

#### Week 1: Telemetry Infrastructure
```typescript
// hooks/useTelemetry.ts
export const useTelemetry = () => {
  const trackAction = useCallback((
    type: UserActionType,
    metadata: ActionMetadata
  ) => {
    const action: UserAction = {
      id: generateId(),
      userId: getCurrentUserId(),
      timestamp: new Date(),
      sessionId: getSessionId(),
      type,
      metadata
    }
    
    // Store locally
    storeTelemetry(action)
    
    // Async aggregate analysis
    updateLocalInsights(action)
  }, [])
  
  return { trackAction }
}
```

#### Week 2: Integration Points
```typescript
// In CardNode component
const { trackAction } = useTelemetry()

const handleCardCreated = (position: XYPosition) => {
  trackAction('card_created', {
    cardPosition: position,
    contentType: 'conversation',
    nearbyCards: findNearbyCards(position, 500),
    canvasZoom: viewport.zoom
  })
}

// In message selection UI
const handleContextSelected = (messages: Message[]) => {
  trackAction('context_selected', {
    selectedMessages: messages.map(m => m.id),
    contextTokenCount: calculateTokens(messages),
    conversationLength: allMessages.length
  })
}
```

#### Week 3: Local Insights Dashboard
Simple analytics view showing:
- Spatial organization patterns (heatmap)
- Most common branching points
- Context selection patterns
- Canvas usage statistics

### Lightweight ML Models (Local-First)

**Model 1: Card Placement Suggester**
```typescript
interface PlacementSuggestion {
  position: { x: number, y: number }
  confidence: number
  reason: 'near_related_content' | 'follows_pattern' | 'optimal_spacing'
}

// Train on user's historical card placements
const suggestCardPlacement = (
  contentType: string,
  relatedCards: Card[]
): PlacementSuggestion => {
  // Analyze user's spatial patterns
  // Return top 3 suggestions
}
```

**Model 2: Context Inheritance Optimizer**
```typescript
interface ContextSuggestion {
  messages: Message[]
  estimatedTokens: number
  reason: string
}

// Learn which message types user typically includes
const suggestContextMessages = (
  conversation: Message[],
  targetBranch: Card
): ContextSuggestion => {
  // Analyze historical context selection patterns
  // Optimize for relevance vs token efficiency
}
```

**Model 3: Merge Candidate Detector**
```typescript
interface MergeCandidate {
  cards: Card[]
  similarity: number
  suggestedSynthesis: string
}

// Identify related conversation threads
const findMergeCandidates = (
  allCards: Card[]
): MergeCandidate[] => {
  // Analyze conversation content similarity
  // Detect thematic connections
}
```

### Training Strategy

**Local-First Approach:**
- Models run entirely in browser (TensorFlow.js or ONNX.js)
- Train on user's own data only
- No server dependency
- Privacy-preserving

**Optional Global Learning:**
- User opt-in for anonymized contribution
- Federated learning approach
- Improve models for all users
- Maintain privacy

### Phase 2 (Data Collection)
- [x] Track 100% of user spatial operations
- [x] Store telemetry in IndexedDB
- [x] Basic insights dashboard functional
- [x] Zero performance impact (<10ms overhead)

### Phase 3 (Intelligence)
- [ ] Card placement suggestions 70%+ accuracy
- [ ] Context selection saves 30%+ tokens on average
- [ ] Merge suggestions accepted 50%+ of time
- [ ] Models run <100ms inference time

### Phase 4 (Moat Validation)
- [ ] Users report tool "knows their thinking"
- [ ] Time-to-organize decreases 40%+ over time
- [ ] Feature becomes differentiator in user interviews

### Privacy & Ethics
- All ML runs locally by default
- Clear user communication about data collection
- Easy opt-out mechanism
- Anonymization for any cloud features

### Performance
- Telemetry collection must be non-blocking
- ML inference budget: <100ms
- Storage quota management (IndexedDB limits)
- Graceful degradation if ML unavailable

### Data Quality
- Handle outliers (accidental card placements)
- Account for experimentation phase (early users)
- Require minimum data threshold before suggestions
- Confidence scores on all predictions

## Why This Creates a Moat

1. **Network Effects**: More usage → better suggestions → more usage
2. **Personalization**: Generic tools can't match user-specific optimization
3. **Compound Advantage**: Gap widens over time as data accumulates
4. **Switching Cost**: Users lose personalized intelligence if they leave

## References & Inspiration

- **Linear**: Learns team workflow patterns for smart defaults
- **Superhuman**: Email triage patterns improve over time
- **Notion AI**: Content pattern recognition for suggestions
- **GitHub Copilot**: Code completion improves with usage

## Next Steps

1. **Immediate**: Implement Phase 2 telemetry infrastructure
2. **Week 2**: Add integration points to existing components
3. **Week 3**: Build basic insights dashboard
4. **Month 2**: Design Phase 3 ML architecture
5. **Month 3**: Ship first intelligent feature (placement suggestions)

---

**The Cheat**: Every user interaction trains your AI
**The Result**: Your product gets smarter daily
**The Moat**: Competitors can't replicate without your data flywheel
